[Security Solution] Add `retryIfConflict` util for `409` conflicts in Integration tests

jpdjere opened the pull request:
## Summary

Fixes: https://github.com/elastic/kibana/issues/171428

**NOTE: the test where this was reported wasn't skipped, so this PR does not unskip any tests.** However, the Flaky Test Runs help us determine that the issue is no longer reproducible.

The `deleteAllPrebuiltRuleAssets` utility reported a `409 Conflict`, presumably from `security-rule` assets that were attempted to be deleted while they were being updated by a parallel process.

This PR wraps the `es.deleteByQuery` calls in the utils `deleteAllPrebuiltRuleAssets` and `deleteAllTimelines` with a new `retryIfConflict` helper, that will retry the operation if the ES request fails with a `409`.

## Flaky test run

`bundled_prebuilt_rules_package` - **ESS** and **Serverless**: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4790

`large_prebuilt_rules_package` - **ESS** and **Serverless**: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4791

`update_prebuilt_rules_package` - **ESS** and **Serverless**: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4792

`management` - **ESS** and **Serverless**: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4793

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.12"]} ONMERGE-->

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

banderror commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/retry_if_conflicts.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+async function waitBeforeNextRetry(): Promise<void> {
+  await new Promise((resolve) => setTimeout(resolve, 200));
```
200 - let's make it the default value of a `retryDelay` parameter.

banderror commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/retry_if_conflicts.ts:
```diff
@@ -0,0 +1,48 @@
+  try {
+    return await operation();
+  } catch (err) {
+    if (!SavedObjectsErrorHelpers.isConflictError(err)) {
+      throw err;
+    }
```
Why are we handling only SO conflict errors here? The 409 error we saw in the test was from ES directly. I think additionally we need to handle ES errors of `"type":"version_conflict_engine_exception"`. Maybe these details don't matter and we could simply check the status code.

jpdjere replied:
See comment in PR

banderror commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/rules/prebuilt_rules/delete_all_prebuilt_rule_assets.ts:
```diff
@@ -4,20 +4,26 @@
+  es: Client,
+  logger: ToolingLog
+): Promise<void> => {
+  await retryIfConflicts(logger, 'deleteAllPrebuiltRuleAssets', async () => {
```
Nit: `'deleteAllPrebuiltRuleAssets'` could probably be replaced by `deleteAllPrebuiltRuleAssets.name` which would be refactoring-safe.

banderror reviewed (commented):
I don't think `retryIfConflicts` does what it should 🙂

jpdjere commented:
@banderror 

Ok, so I had to rethink this a bit. 

Responses from the Elasticsearch client vary from one type of operation to another, so there's no `status` or `statusCode` property that I can read from all different types of ES responses to check if they are `409`. The response from `DeleteByQuery` looks like:
```ts
export interface DeleteByQueryResponse {
    batches?: long;
    deleted?: long;
    failures?: BulkIndexByScrollFailure[];
    noops?: long;
    requests_per_second?: float;
    // other stuff......
    total?: long;
    version_conflicts?: long;
}
```
and as you can see from the error in the ticket, the `409` error is reported in the `failures` array of the request body. And this `failure` prop doesn't exist in other ES responses.

(Also, I cannot get access to anything else apart from the body when doing requests directly to ES in integration tests)

So I refactored the code so that this utility is specific to retry DeleteByQuery operations if it hits 409 conflict errors, which seems to be, at least until now, the ES operation where this conflict is happening.

I think we can try to generalize this utility, or create other specific version of retries if we need them in the future. Let me know what you think.

banderror commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/utils/retry_delete_by_query_conflicts.ts:
```diff
@@ -0,0 +1,53 @@
+import { DeleteByQueryResponse } from '@elastic/elasticsearch/lib/api/types';
+import { ToolingLog } from '@kbn/tooling-log';
+
+// Number of times to retry when conflicts occur
+export const RetryForConflictsAttempts = 2;
+
+// Delay between retries when conflicts occur
+export const RetryForConflictsDelay = 200;
```
Nit: Normally we name constant literals `THIS_WAY`.

jpdjere replied:
Yes, thanks, renaming.

banderror reviewed (approved):
The Flaky Test Runner is happy. Approving the PR to not being a blocker for it - posted only a bunch of minor comments. Thanks for addressing the feedback, LGTM 👍

jpdjere commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/default_license/prebuilt_rules/large_prebuilt_rules_package/install_large_prebuilt_rules_package.ts:
```diff
@@ -21,7 +21,8 @@ export default ({ getService }: FtrProviderContext): void => {
       await deleteAllRules(supertest, log);
-      await deleteAllPrebuiltRuleAssets(es, log);
+      const test = await deleteAllPrebuiltRuleAssets(es, log);
+      console.log('\n\n\nHola a todos: ', test, '\n\n\n', '\n\n\n');
```
Not sure how that showed up for you 🤔 

I pushed my final changes before you reviewed. But anyways, it's not there now.

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.12|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 174185
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

jpdjere commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.12|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/174762">](https://github.com/elastic/kibana/pull/174762)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.8 -->

kibanamachine commented:
Looks like this PR has a backport PR but it still hasn't been merged. Please merge it ASAP to keep the branches relatively in sync.

mistic commented:
This PR didn't make it into the latest BC for 8.12.0. Updating the labels.
